native/src/main/native/org/mozilla/jss/pkcs11/PK11KeyGenerator.c

```diff
@@ -282,53 +283,61 @@ Java_org_mozilla_jss_pkcs11_PK11KeyGenerator_generatePBE(
     }
     /* print_secitem(pwitem); */
 
-    mech = JSS_getPK11MechFromAlg(env, alg);
-
-    if( mech == CKM_PBA_SHA1_WITH_SHA1_HMAC ) {
+    if(hashAlg==NULL) {
+    	mech = JSS_getPK11MechFromAlg(env, alg);
+    }
 
+    switch(mech){
+    case CKM_PBA_SHA1_WITH_SHA1_HMAC:
         /* special case, construct key by hand. Bug #336587 */
-
         skey = constructSHA1PBAKey(env, slot, pwitem, salt, iterationCount);
-        if( skey==NULL ) {
-            /* exception was thrown */
-            goto finish;
-        }
-
-    } else {
-
-        /* get the algorithm info */
-        oidTag = JSS_getOidTagFromAlg(env, alg);
-        PR_ASSERT(oidTag != SEC_OID_UNKNOWN);
-
-        SECOidTag encAlgOidTag = JSS_getOidTagFromAlg(env, encAlg);
-        PR_ASSERT(encAlgOidTag != SEC_OID_UNKNOWN);
-
-        /* create algid */
-        algid = PK11_CreatePBEV2AlgorithmID(
-            oidTag,
-            encAlgOidTag,
-            SEC_OID_HMAC_SHA1,
-            0,
-            iterationCount,
-            salt);
-
-        if( algid == NULL ) {
-            JSS_throwMsg(env, TOKEN_EXCEPTION,
-                    "Unable to process PBE parameters");
-            goto finish;
-        }
-
-        /* generate the key */
-        skey = PK11_PBEKeyGen(slot, algid, pwitem, PR_FALSE /*faulty3DES*/,
-                        NULL /*wincx*/);
-        if( skey == NULL ) {
-            JSS_throwMsg(env, TOKEN_EXCEPTION, "Failed to generate PBE key");
-            goto finish;
+        break;
```

Comment:
OK, after looking at the original file, I think I see what this is doing. Previously, the code was just an (if) for SHA1 amd (else) for everything else, which results in using a PBEV2 algid. You have just made this a switch where you've added a case for SEC_OID_SHA256, which takes it's own action.

My question here, just for my own understanding, is what case are we satisfying here? Is this something you ran into when experimenting with various p12 files and found out you needed this? Also should there be 384 and 512 variants.   Trying to refresh my memory, I think the case we were trying to debug was ending up in the (default) clause. I'm just trying to get an understanding where the added case comes into play and if you saw this code being hit? Thanks.

Reply:
Finally, I have reverted the code to use the `if`. However, I have add a new `if`  for  SHA-[256, 384, 512]. In case the symmetric key is not generated a last attempt is done with the default case. To note that the `skey` is null only if the other `if` are not used because if they are and the key is `NULL` the execution go the `finish` label at the end of the method.

Additionally, I have add a new method for the new case to avoid code repetition.